open-craft/openedx-completion-aggregator
# [MCKIN-7831] Make single-user aggregator endpoint calculate on the fly

**Description:** Describe in a couple of sentence what this PR adds

**JIRA:** MCKIN-7831, OC-4787

**Dependencies:** None

**Merge deadline:** None

**Installation instructions:** None

**Testing instructions:**

1. Ensure the `completion.enable_completion_tracking` waffle switch is enabled in the lms, and 
    `settings.COMPLETION_AGGREGATOR_ASYNC_AGGREGATION = True`.
1. Create some BlockCompletions, either by browsing a course, or from a `./manage.py lms shell`
1. Verify (from the logging or from the command line) that `StaleCompletions` are being created (but not resolved), and the corresponding `Aggregator` objects are not created/updated.
1. Using a staff account, hit the multi-user API endpoint, and verify (from the logging or from the command line) that:
   1. `StaleCompletions` are not being resolved, and
   1. `Aggregations` are not created/recalculated before they are returned to the API.

   Multi-user API endpoints are (as `staff` user`):
   ```
   GET /api/completion-aggregator/v1/course/<course_key>/
   ```

1. Hit the versioned single-user API endpoint(s) for each course, and verify (from the logging or from the command line) that:
   1. `StaleCompletions` are being resolved, and
   1. `Aggregations` are created/recalculated before they are returned to the API.
   
   single-user API endpoints to test are (as non-staff):
   ```
   GET /api/completion-aggregator/v0/course/
   GET /api/completion-aggregator/v0/course/<course_key>/
   GET /api/completion-aggregator/v1/course/
   GET /api/completion-aggregator/v1/course/<course_key>/
   ```
   or as staff:
   ```
   GET /api/completion-aggregator/v1/course/<course_key>/?username=<user_name>
   ```

**Reviewers:**

**Merge checklist:**

**Post merge:**
      finished.

**Author concerns:** List any concerns about this PR - inelegant 
solutions, hacks, quick-and-dirty implementations, concerns about 
migrations, etc.
